**Dplyr** is one of the most popular r-packages and also part of **tidyverse** that's been developed by Hadley Wickham. The mere fact that dplyr package is very famous means, it's one of the most frequently used. Being a data scientist is not always about creating sophisticated models but Data Analysis (Manipulation) and Data Visualization play a very important role in BAU of many us - in fact, a very important part before any modeling exercise since Feature Engineering and EDA are the most important differentiating factors of your model and someone else's.

Hence, this notebook aims to bring out some well known and not-so-well-known applications of dplyr so that any data analyst could leverage its potential

In [1]:
# This R environment comes with all of CRAN preinstalled, as well as many other helpful packages
# The environment is defined by the kaggle/rstats docker image: https://github.com/kaggle/docker-rstats
# For example, here's several helpful packages to load in 

library(dplyr) # Loading Dplyr package

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list the files in the input directory

system("ls ../input")

# Any results you write to the current directory are saved as output.


Attaching package: ‘dplyr’

The following objects are masked from ‘package:stats’:

    filter, lag

The following objects are masked from ‘package:base’:

    intersect, setdiff, setequal, union



Let us start by reading the input training file using the base r function `read.csv`

In [2]:
train <- read.csv('../input/train.csv',stringsAsFactors = F, header = T)
train #just to see if it's been loaded

PassengerId Survived Pclass
1    1          0        3     
2    2          1        1     
3    3          1        3     
4    4          1        1     
5    5          0        3     
6    6          0        3     
7    7          0        1     
8    8          0        3     
9    9          1        3     
10  10          1        2     
11  11          1        3     
12  12          1        1     
13  13          0        3     
14  14          0        3     
15  15          0        3     
16  16          1        2     
17  17          0        3     
18  18          1        2     
19  19          0        3     
20  20          1        3     
21  21          0        2     
22  22          1        2     
23  23          1        3     
24  24          1        1     
25  25          0        3     
26  26          1        3     
27  27          0        3     
28  28          0        1     
29  29          1        3     
30  30          0        3     
⋮   ⋮           ⋮        ⋮     
862 862         0        2     
863 863         1        1     
864 864         0        3     
865 865         0        2     
866 866         1        2     
867 867         1        2     
868 868         0        1     
869 869         0        3     
870 870         1        3     
871 871         0        3     
872 872         1        1     
873 873         0        1     
874 874         0        3     
875 875         1        2     
876 876         1        3     
877 877         0        3     
878 878         0        3     
879 879         0        3     
880 880         1        1     
881 881         1        2     
882 882         0        3     
883 883         0        3     
884 884         0        2     
885 885         0        3     
886 886         0        3     
887 887         0        2     
888 888         1        1     
889 889         0        3     
890 890         1        1     
891 891         0        3     
    Name                                                      Sex    Age SibSp
1   Braund, Mr. Owen Harris                                   male   22  1    
2   Cumings, Mrs. John Bradley (Florence Briggs Thayer)       female 38  1    
3   Heikkinen, Miss. Laina                                    female 26  0    
4   Futrelle, Mrs. Jacques Heath (Lily May Peel)              female 35  1    
5   Allen, Mr. William Henry                                  male   35  0    
6   Moran, Mr. James                                          male   NA  0    
7   McCarthy, Mr. Timothy J                                   male   54  0    
8   Palsson, Master. Gosta Leonard                            male    2  3    
9   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         female 27  0    
10  Nasser, Mrs. Nicholas (Adele Achem)                       female 14  1    
11  Sandstrom, Miss. Marguerite Rut                           female  4  1    
12  Bonnell, Miss. Elizabeth                                  female 58  0    
13  Saundercock, Mr. William Henry                            male   20  0    
14  Andersson, Mr. Anders Johan                               male   39  1    
15  Vestrom, Miss. Hulda Amanda Adolfina                      female 14  0    
16  Hewlett, Mrs. (Mary D Kingcome)                           female 55  0    
17  Rice, Master. Eugene                                      male    2  4    
18  Williams, Mr. Charles Eugene                              male   NA  0    
19  Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)   female 31  1    
20  Masselmani, Mrs. Fatima                                   female NA  0    
21  Fynney, Mr. Joseph J                                      male   35  0    
22  Beesley, Mr. Lawrence                                     male   34  0    
23  McGowan, Miss. Anna "Annie"                               female 15  0    
24  Sloper, Mr. William Thompson                              male   28  0    
25  Palsson, Miss. Torborg Danira            

Getting the total number of rows in the given dataframe (even though it's been very straight forward with nrow() in base-r, this being a dplyr starter-kit, we'll start with that.

In [3]:
train %>% count()

n  
1 891

The above code just gives the row count of the dataframe that's been passed with the pipe %>% operator. The pipe operator works very similar to the | (pipe) operator in Unix environment where the ouput of the current operation is fed as the input of the following operation. Similary in dplyr or any other package that supports pipe operator, the functions in it will always take only dataframe as the first arugment hence the function can be called in two ways like below:

In [4]:
count(train) #Without pipe, passing the df as the first argument
train %>% count() #with pipe, more convient and more readability

n  
1 891

n  
1 891

But dplyr's real flavor starts with the following 5 functions (or as most people call, verbs of dplyr):

* select()
* filter()	
* arrange() 
* mutate()	
* summarise()	
* group_by()	

And let us see what every one of these does!

### select:

select() as the name suggests selects the columns that are required from a given dataframe and if multiple columns are required or not required, then `one_of()` could be used within select. 

In [5]:
select(train,Age) #without pipe 

Age
1   22 
2   38 
3   26 
4   35 
5   35 
6   NA 
7   54 
8    2 
9   27 
10  14 
11   4 
12  58 
13  20 
14  39 
15  14 
16  55 
17   2 
18  NA 
19  31 
20  NA 
21  35 
22  34 
23  15 
24  28 
25   8 
26  38 
27  NA 
28  19 
29  NA 
30  NA 
⋮   ⋮  
862 21 
863 48 
864 NA 
865 24 
866 42 
867 27 
868 31 
869 NA 
870  4 
871 26 
872 47 
873 33 
874 47 
875 28 
876 15 
877 20 
878 19 
879 NA 
880 56 
881 25 
882 33 
883 22 
884 28 
885 25 
886 39 
887 27 
888 19 
889 NA 
890 26 
891 32

In [6]:
#multicolumn selection
train %>% select(one_of('Sex','Age')) 

Sex    Age
1   male   22 
2   female 38 
3   female 26 
4   female 35 
5   male   35 
6   male   NA 
7   male   54 
8   male    2 
9   female 27 
10  female 14 
11  female  4 
12  female 58 
13  male   20 
14  male   39 
15  female 14 
16  female 55 
17  male    2 
18  male   NA 
19  female 31 
20  female NA 
21  male   35 
22  male   34 
23  female 15 
24  male   28 
25  female  8 
26  female 38 
27  male   NA 
28  male   19 
29  female NA 
30  male   NA 
⋮   ⋮      ⋮  
862 male   21 
863 female 48 
864 female NA 
865 male   24 
866 female 42 
867 female 27 
868 male   31 
869 male   NA 
870 male    4 
871 male   26 
872 female 47 
873 male   33 
874 male   47 
875 female 28 
876 female 15 
877 male   20 
878 male   19 
879 male   NA 
880 female 56 
881 female 25 
882 male   33 
883 female 22 
884 male   28 
885 male   25 
886 female 39 
887 male   27 
888 female 19 
889 female NA 
890 male   26 
891 male   32

In [7]:
#multicolumn rejection

train %>% select(-one_of('Age','Sex'))

PassengerId Survived Pclass
1    1          0        3     
2    2          1        1     
3    3          1        3     
4    4          1        1     
5    5          0        3     
6    6          0        3     
7    7          0        1     
8    8          0        3     
9    9          1        3     
10  10          1        2     
11  11          1        3     
12  12          1        1     
13  13          0        3     
14  14          0        3     
15  15          0        3     
16  16          1        2     
17  17          0        3     
18  18          1        2     
19  19          0        3     
20  20          1        3     
21  21          0        2     
22  22          1        2     
23  23          1        3     
24  24          1        1     
25  25          0        3     
26  26          1        3     
27  27          0        3     
28  28          0        1     
29  29          1        3     
30  30          0        3     
⋮   ⋮           ⋮        ⋮     
862 862         0        2     
863 863         1        1     
864 864         0        3     
865 865         0        2     
866 866         1        2     
867 867         1        2     
868 868         0        1     
869 869         0        3     
870 870         1        3     
871 871         0        3     
872 872         1        1     
873 873         0        1     
874 874         0        3     
875 875         1        2     
876 876         1        3     
877 877         0        3     
878 878         0        3     
879 879         0        3     
880 880         1        1     
881 881         1        2     
882 882         0        3     
883 883         0        3     
884 884         0        2     
885 885         0        3     
886 886         0        3     
887 887         0        2     
888 888         1        1     
889 889         0        3     
890 890         1        1     
891 891         0        3     
    Name                                                      SibSp Parch
1   Braund, Mr. Owen Harris                                   1     0    
2   Cumings, Mrs. John Bradley (Florence Briggs Thayer)       1     0    
3   Heikkinen, Miss. Laina                                    0     0    
4   Futrelle, Mrs. Jacques Heath (Lily May Peel)              1     0    
5   Allen, Mr. William Henry                                  0     0    
6   Moran, Mr. James                                          0     0    
7   McCarthy, Mr. Timothy J                                   0     0    
8   Palsson, Master. Gosta Leonard                            3     1    
9   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         0     2    
10  Nasser, Mrs. Nicholas (Adele Achem)                       1     0    
11  Sandstrom, Miss. Marguerite Rut                           1     1    
12  Bonnell, Miss. Elizabeth                                  0     0    
13  Saundercock, Mr. William Henry                            0     0    
14  Andersson, Mr. Anders Johan                               1     5    
15  Vestrom, Miss. Hulda Amanda Adolfina                      0     0    
16  Hewlett, Mrs. (Mary D Kingcome)                           0     0    
17  Rice, Master. Eugene                                      4     1    
18  Williams, Mr. Charles Eugene                              0     0    
19  Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)   1     0    
20  Masselmani, Mrs. Fatima                                   0     0    
21  Fynney, Mr. Joseph J                                      0     0    
22  Beesley, Mr. Lawrence                                     0     0    
23  McGowan, Miss. Anna "Annie"                               0     0    
24  Sloper, Mr. William Thompson                              0     0    
25  Palsson, Miss. Torborg Danira                             3     1    
26  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) 1     5    
27  Emir, Mr. Farred C

Like selecting  a column with entire column name (or multiple column names with one_of()), select could also be used with a few more string ops. 

In [8]:
train %>% select(starts_with('P'))

PassengerId Pclass Parch
1    1          3      0    
2    2          1      0    
3    3          3      0    
4    4          1      0    
5    5          3      0    
6    6          3      0    
7    7          1      0    
8    8          3      1    
9    9          3      2    
10  10          2      0    
11  11          3      1    
12  12          1      0    
13  13          3      0    
14  14          3      5    
15  15          3      0    
16  16          2      0    
17  17          3      1    
18  18          2      0    
19  19          3      0    
20  20          3      0    
21  21          2      0    
22  22          2      0    
23  23          3      0    
24  24          1      0    
25  25          3      1    
26  26          3      5    
27  27          3      0    
28  28          1      2    
29  29          3      0    
30  30          3      0    
⋮   ⋮           ⋮      ⋮    
862 862         2      0    
863 863         1      0    
864 864         3      2    
865 865         2      0    
866 866         2      0    
867 867         2      0    
868 868         1      0    
869 869         3      0    
870 870         3      1    
871 871         3      0    
872 872         1      1    
873 873         1      0    
874 874         3      0    
875 875         2      0    
876 876         3      0    
877 877         3      0    
878 878         3      0    
879 879         3      0    
880 880         1      1    
881 881         2      1    
882 882         3      0    
883 883         3      0    
884 884         2      0    
885 885         3      0    
886 886         3      5    
887 887         2      0    
888 888         1      0    
889 889         3      2    
890 890         1      0    
891 891         3      0

In [9]:
train %>% select(ends_with('e'))

Name                                                      Age Fare    
1   Braund, Mr. Owen Harris                                   22    7.2500
2   Cumings, Mrs. John Bradley (Florence Briggs Thayer)       38   71.2833
3   Heikkinen, Miss. Laina                                    26    7.9250
4   Futrelle, Mrs. Jacques Heath (Lily May Peel)              35   53.1000
5   Allen, Mr. William Henry                                  35    8.0500
6   Moran, Mr. James                                          NA    8.4583
7   McCarthy, Mr. Timothy J                                   54   51.8625
8   Palsson, Master. Gosta Leonard                             2   21.0750
9   Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)         27   11.1333
10  Nasser, Mrs. Nicholas (Adele Achem)                       14   30.0708
11  Sandstrom, Miss. Marguerite Rut                            4   16.7000
12  Bonnell, Miss. Elizabeth                                  58   26.5500
13  Saundercock, Mr. William Henry                            20    8.0500
14  Andersson, Mr. Anders Johan                               39   31.2750
15  Vestrom, Miss. Hulda Amanda Adolfina                      14    7.8542
16  Hewlett, Mrs. (Mary D Kingcome)                           55   16.0000
17  Rice, Master. Eugene                                       2   29.1250
18  Williams, Mr. Charles Eugene                              NA   13.0000
19  Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)   31   18.0000
20  Masselmani, Mrs. Fatima                                   NA    7.2250
21  Fynney, Mr. Joseph J                                      35   26.0000
22  Beesley, Mr. Lawrence                                     34   13.0000
23  McGowan, Miss. Anna "Annie"                               15    8.0292
24  Sloper, Mr. William Thompson                              28   35.5000
25  Palsson, Miss. Torborg Danira                              8   21.0750
26  Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson) 38   31.3875
27  Emir, Mr. Farred Chehab                                   NA    7.2250
28  Fortune, Mr. Charles Alexander                            19  263.0000
29  O'Dwyer, Miss. Ellen "Nellie"                             NA    7.8792
30  Todoroff, Mr. Lalio                                       NA    7.8958
⋮   ⋮                                                         ⋮   ⋮       
862 Giles, Mr. Frederick Edward                               21  11.5000 
863 Swift, Mrs. Frederick Joel (Margaret Welles Barron)       48  25.9292 
864 Sage, Miss. Dorothy Edith "Dolly"                         NA  69.5500 
865 Gill, Mr. John William                                    24  13.0000 
866 Bystrom, Mrs. (Karolina)                                  42  13.0000 
867 Duran y More, Miss. Asuncion                              27  13.8583 
868 Roebling, Mr. Washington Augustus II                      31  50.4958 
869 van Melkebeke, Mr. Philemon                               NA   9.5000 
870 Johnson, Master. Harold Theodor                            4  11.1333 
871 Balkic, Mr. Cerin                                         26   7.8958 
872 Beckwith, Mrs. Richard Leonard (Sallie Monypeny)          47  52.5542 
873 Carlsson, Mr. Frans Olof                                  33   5.0000 
874 Vander Cruyssen, Mr. Victor                               47   9.0000 
875 Abelson, Mrs. Samuel (Hannah Wizosky)                     28  24.0000 
876 Najib, Miss. Adele Kiamie "Jane"                          15   7.2250 
877 Gustafsson, Mr. Alfred Ossian                             20   9.8458 
878 Petroff, Mr. Nedelio                                      19   7.8958 
879 Laleff, Mr. Kristo                                        NA   7.8958 
880 Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)             56  83.1583 
881 Shelley, Mrs. William (Imanita Parrish Hall)              25  26.0000 
882 Markun, Mr. Johann                                        33   7.8958 
883 Dahlberg, Miss. Gerda Ulr

### Group_by:

Group_by is a lot similar to SQL Group by but more versatile. It is related to concept of “split-apply-combine”. Let us understand group_by with a starter example of finding out number of male and number of female - which logically could be the count of each Sex Type (once grouped by Sex).



In [10]:
train %>% group_by(Sex) %>% count()

Sex    n  
1 female 314
2 male   577

Aha! That seems simple and now let us do a two level grouping to understand how many of survived of each gender.

In [11]:
train %>% group_by(Survived, Sex) %>% count()

train %>% group_by(Sex, Survived) %>% count()

Survived Sex    n  
1 0        female  81
2 0        male   468
3 1        female 233
4 1        male   109

Sex    Survived n  
1 female 0         81
2 female 1        233
3 male   0        468
4 male   1        109

### Mutate and Summarise:

That's minimally group_by, but the true power of group_by is unveiled only when it is coupled with mutate and summarise functions.

Mutate function adds a new column based on the given expression while summarise function summarises the dataset based on the given function and let us see the difference in action with the following example.

Let us get the average age of all survivors (and non-survivors): so this must be group_by -ed based on Survived while summarised by Age so that we will get a summarised mean value.for two groups.

In [12]:
train %>% group_by(Survived) %>% summarise(mean(Age))

 #Remember we have got NAs, so mean() wouldn't work and to bypass NAs, na.rm = T must be passed. 

train %>% group_by(Survived) %>% summarise(average_age = mean(Age,na.rm=T))

Survived mean(Age)
1 0        NA       
2 1        NA

Survived average_age
1 0        30.62618   
2 1        28.34369

That's summarise() giving us the summary of the dataframe. If we need to create a new column, values filled for all 891 datapoints, that's where mutate plays its role. Let us create a new column, `Age_Bracket` containing value `Minor` if Age is less than 18 else `Major`

In [13]:
train %>% mutate(Age_Bracket = ifelse(Age < 18, 'Minor','Major')) %>% select(starts_with('Age'))

#In fact this can be coupled with Survivor list to see the impact of this Age_bracket

train %>% 
mutate(Age_Bracket = ifelse(Age < 18, 'Minor','Major')) %>% 
group_by(Survived,Age_Bracket) %>% 
summarise(pnt = (n()/nrow(train))*100)


Age Age_Bracket
1   22  Major      
2   38  Major      
3   26  Major      
4   35  Major      
5   35  Major      
6   NA  NA         
7   54  Major      
8    2  Minor      
9   27  Major      
10  14  Minor      
11   4  Minor      
12  58  Major      
13  20  Major      
14  39  Major      
15  14  Minor      
16  55  Major      
17   2  Minor      
18  NA  NA         
19  31  Major      
20  NA  NA         
21  35  Major      
22  34  Major      
23  15  Minor      
24  28  Major      
25   8  Minor      
26  38  Major      
27  NA  NA         
28  19  Major      
29  NA  NA         
30  NA  NA         
⋮   ⋮   ⋮          
862 21  Major      
863 48  Major      
864 NA  NA         
865 24  Major      
866 42  Major      
867 27  Major      
868 31  Major      
869 NA  NA         
870  4  Minor      
871 26  Major      
872 47  Major      
873 33  Major      
874 47  Major      
875 28  Major      
876 15  Minor      
877 20  Major      
878 19  Major      
879 NA  NA         
880 56  Major      
881 25  Major      
882 33  Major      
883 22  Major      
884 28  Major      
885 25  Major      
886 39  Major      
887 27  Major      
888 19  Major      
889 NA  NA         
890 26  Major      
891 32  Major

Survived Age_Bracket pnt      
1 0        Major       41.750842
2 0        Minor        5.836139
3 0        NA          14.029181
4 1        Major       25.701459
5 1        Minor        6.846240
6 1        NA           5.836139

That's how dplyr can get more powerful with group_by coupled with mutate or summarise for feautre engineering and for better data visualization. But this doesn't stop here, because one of the most important function a dataanalyst would require is sorting and that's what `arrange()` does.

## arrange - in ascending order:

In [14]:
train %>% arrange(Fare) %>% tail(22) #Extracting last 22 results after sorting the fare in asending order

PassengerId Survived Pclass
870 319         1        1     
871 857         1        1     
872 690         1        1     
873 731         1        1     
874 780         1        1     
875 378         0        1     
876 528         0        1     
877 381         1        1     
878 558         0        1     
879 701         1        1     
880 717         1        1     
881 119         0        1     
882 300         1        1     
883 312         1        1     
884 743         1        1     
885  28         0        1     
886  89         1        1     
887 342         1        1     
888 439         0        1     
889 259         1        1     
890 680         1        1     
891 738         1        1     
    Name                                                  Sex    Age SibSp
870 Wick, Miss. Mary Natalie                              female 31  0    
871 Wick, Mrs. George Dennick (Mary Hitchcock)            female 45  1    
872 Madill, Miss. Georgette Alexandra                     female 15  0    
873 Allen, Miss. Elisabeth Walton                         female 29  0    
874 Robert, Mrs. Edward Scott (Elisabeth Walton McMillan) female 43  0    
875 Widener, Mr. Harry Elkins                             male   27  0    
876 Farthing, Mr. John                                    male   NA  0    
877 Bidois, Miss. Rosalie                                 female 42  0    
878 Robbins, Mr. Victor                                   male   NA  0    
879 Astor, Mrs. John Jacob (Madeleine Talmadge Force)     female 18  1    
880 Endres, Miss. Caroline Louise                         female 38  0    
881 Baxter, Mr. Quigg Edmond                              male   24  0    
882 Baxter, Mrs. James (Helene DeLaudeniere Chaput)       female 50  0    
883 Ryerson, Miss. Emily Borie                            female 18  2    
884 Ryerson, Miss. Susan Parker "Suzette"                 female 21  2    
885 Fortune, Mr. Charles Alexander                        male   19  3    
886 Fortune, Miss. Mabel Helen                            female 23  3    
887 Fortune, Miss. Alice Elizabeth                        female 24  3    
888 Fortune, Mr. Mark                                     male   64  1    
889 Ward, Miss. Anna                                      female 35  0    
890 Cardeza, Mr. Thomas Drake Martinez                    male   36  0    
891 Lesurer, Mr. Gustave J                                male   35  0    
    Parch Ticket   Fare     Cabin           Embarked
870 2     36928    164.8667 C7              S       
871 1     36928    164.8667                 S       
872 1     24160    211.3375 B5              S       
873 0     24160    211.3375 B5              S       
874 1     24160    211.3375 B3              S       
875 2     113503   211.5000 C82             C       
876 0     PC 17483 221.7792 C95             S       
877 0     PC 17757 227.5250                 C       
878 0     PC 17757 227.5250                 C       
879 0     PC 17757 227.5250 C62 C64         C       
880 0     PC 17757 227.5250 C45             C       
881 1     PC 17558 247.5208 B58 B60         C       
882 1     PC 17558 247.5208 B58 B60         C       
883 2     PC 17608 262.3750 B57 B59 B63 B66 C       
884 2     PC 17608 262.3750 B57 B59 B63 B66 C       
885 2     19950    263.0000 C23 C25 C27     S       
886 2     19950    263.0000 C23 C25 C27     S       
887 2     19950    263.0000 C23 C25 C27     S       
888 4     19950    263.0000 C23 C25 C27     S       
889 0     PC 17755 512.3292                 C       
890 1     PC 17755 512.3292 B51 B53 B55     C       
891 0     PC 17755 512.3292 B101            C

In [15]:
## Arrange in descending order

train %>% arrange(desc(Age)) %>% head(10)

PassengerId Survived Pclass Name                                 Sex  Age 
1  631         1        1      Barkworth, Mr. Algernon Henry Wilson male 80.0
2  852         0        3      Svensson, Mr. Johan                  male 74.0
3   97         0        1      Goldschmidt, Mr. George B            male 71.0
4  494         0        1      Artagaveytia, Mr. Ramon              male 71.0
5  117         0        3      Connors, Mr. Patrick                 male 70.5
6  673         0        2      Mitchell, Mr. Henry Michael          male 70.0
7  746         0        1      Crosby, Capt. Edward Gifford         male 70.0
8   34         0        2      Wheadon, Mr. Edward H                male 66.0
9   55         0        1      Ostby, Mr. Engelhart Cornelius       male 65.0
10 281         0        3      Duane, Mr. Frank                     male 65.0
   SibSp Parch Ticket     Fare    Cabin Embarked
1  0     0     27042      30.0000 A23   S       
2  0     0     347060      7.7750       S       
3  0     0     PC 17754   34.6542 A5    C       
4  0     0     PC 17609   49.5042       C       
5  0     0     370369      7.7500       Q       
6  0     0     C.A. 24580 10.5000       S       
7  1     1     WE/P 5735  71.0000 B22   S       
8  0     0     C.A. 24579 10.5000       S       
9  0     1     113509     61.9792 B30   C       
10 0     0     336439      7.7500       Q

### filter:

Filter does row_wise filter ( similar to what select did with columns). filter() takes a logical expression and evaluates them and results the only_true datapoints. So to be clear, all that matters to filter() function is if the expression evaluates to TRUE.

Let us start with filtering (extracting) only male and getting their Embarked station count.  

In [16]:
train %>% 
filter(Sex == 'male') %>%
group_by(Embarked) %>%
count()

Embarked n  
1 C         95
2 Q         41
3 S        441

In [17]:
#Getting the count of everyone whose age is lesser than 18

train %>% filter(Age < 18) %>% count()

n  
1 113

Coupling Filter with Regex to perform simple string manipulation and detection.

In [18]:
train %>% filter(grepl('wick',train$Name))

PassengerId Survived Pclass Name                                       Sex 
1 344         0        2      Sedgwick, Mr. Charles Frederick Waddington male
  Age SibSp Parch Ticket Fare Cabin Embarked
1 25  0     0     244361 13         S

*And this is dplyr in a nut shell and hope you get a decent start with this notebook if you are a beginner. Please share your thoughts in comments and suggestions! 
*
